x-pack/plugins/security/server/authentication/index.ts

```diff
@@ -105,28 +105,30 @@ export async function setupAuthentication({
       // authentication (username and password) or arbitrary external page managed by 3rd party
       // Identity Provider for SSO authentication mechanisms. Authentication provider is the one who
       // decides what location user should be redirected to.
-      return t.redirected(authenticationResult.redirectURL!);
+      return response.redirected(undefined, {
+        headers: {
+          location: authenticationResult.redirectURL!,
+        },
+      });
     }
 
-    let error;
     if (authenticationResult.failed()) {
       authLogger.info(`Authentication attempt failed: ${authenticationResult.error!.message}`);
-      error = wrapError(authenticationResult.error);
-
-      const authResponseHeaders = authenticationResult.authResponseHeaders;
-      for (const [headerName, headerValue] of Object.entries(authResponseHeaders || {})) {
-        if (error.output.headers[headerName] !== undefined) {
-          authLogger.warn(`Server rewrites a error response header [${headerName}].`);
-        }
-        // Hapi typings don't support headers that are `string[]`.
-        error.output.headers[headerName] = headerValue as any;
-      }
+      const error = authenticationResult.error;
```

Comment:
__note:__ if the change in behavior above is okay (`internalError` for unexpected errors), here it's a bit different, ideally we should return the same status Elasticsearch gives us during authentication, e.g. `500` or whatever ES may return in the future. Ideally `401` should only be returned when ES returns error with such status or when we can't handle authentication (`notHandled`). It feels like we need something like `wrapError` in the core/responseFactory....

Reply:
yes, that question I asked here https://github.com/elastic/kibana/pull/42442#discussion_r309710069 ok, I will add customizable error handler